sealir-tutorials/ch09_whole_program_compiler_driver.py

```diff
@@ -0,0 +1,326 @@
+#
+# Lastly, the function `get_call_graph` returns the call graph.
+
+class CallGraphVisitor(ast.NodeVisitor):
+
+    def __init__(self, source_code, file_name):
+        # Stash the arguments
+        self.source_code = source_code
+        self.file_name = file_name
+        # Get the AST once
+        self.tree = ast.parse(source_code)
+        # Initialize the cpython symtable
+        self.symt = symtable.symtable(source_code, file_name , "exec")
+        # Filter out all class definitions from the AST
+        self.classes = set((node.name for node in ast.walk(self.tree) if
+                        isinstance(node, ast.ClassDef)))
+        # Setup the namespace and class stacks
+        self.namespace_stack = []
+        self.class_stack = []
+        # Nested dictionary to record class types
+        self.class_types = defaultdict(dict)
+        # Dictionary to record all functions, this is effectively the symbol
+        # table.
+        self.functions = {}
+        # List of all global ast.Call nodes
+        self.global_calls = []
+
+    def get_call_graph(self) -> dict[str: tuple[str]]:
+        """Obtain a call graph suitable for processing with networkx.
+
+        Returns a dictionary mapping function names as strings to lists of
+        function names as strings.
+        """
+
+        return {k:tuple(c[1] for c in v.calls) for k,v in self.functions.items()}
+
+
+    def update_calls(self, node):
+        """Update the calls for a function or register a global call."""
+        # Flatten the name of the call from ast.Attribute or ast.Name
+        call_qname = attribute_to_qualified_name(node)
+        class_name = self.class_stack[-1] if self.class_stack else None
+        if call_qname.startswith("self"):
+            # If the call starts with "self", it is a method call, we replace
+            # the "self" with the current class name to qualify it.
+            call_qname = class_name + call_qname[4:]
+        if class_name and call_qname.startswith(class_name):
+            # Replace calls from class attributes with their qualified name.
+            # First split the qualified name by the dot separator.
+            split_qname = call_qname.split(".")
+            # Get the types of the current classes attributes.
+            current_class_types = self.class_types[class_name]
+            # If the second element in the qualified name matches the name of
+            # the class attribute, replace the reference to the class.attribute
+            # string with the correct type.
+            if split_qname[1] in current_class_types:
+                call_qname = ".".join([current_class_types[split_qname[1]]] +
+                                      split_qname[2:])
+        if call_qname in self.classes:
+            # If the call ends with the current class name, we replace it with
+            # the constructor call, since this is the Python semantics.
+            call_qname = call_qname + ".__init__"
```

Comment:
This is good enough for now.

In the future, this can be quite complicated; e.g. `__new__` and metaclass stuffs.

Reply:
yes